### PR TITLE
adding file to fine tune model

### LINKED ISSUE #17: Fine-Tuning the CodeT5 Model for R Language in Bioinformatics
The CodeT5 model needs to be fine-tuned using bioinformatics-specific R datasets to enhance its code generation, debugging, and optimization capabilities. The fine-tuned model should understand Bioconductor functions commonly used in bioinformatics research.

Acceptance Criteria:
- Preprocess the collected dataset for fine-tuning.
- Fine-tune the CodeT5 model using the curated R language datasets.

### PR DESCRIPTION
**Pull request by Omp06**
Fixes #17 

**What was changed?**

The CodeT5 model has the ability to be fine-tuned on a new dataset and task. I added the files needed to be able to run the fine-tuning process on the task of our choice. I also added the file needed to load and then tune the model to our liking. These files will be run using GPU3 in Hopper where we will have the power and space needed for finetuning. 

**Why was it changed?**

The CodeT5 model does not have the ability to generate and refine R code that bioinformatics researchers use. The run_gen.py file allows us to fine-tune the model to be used for the new tasks that we require. The models.py file holds functions that support the fine-tuning process such as building and loading the model itself. 

**How was it changed?**

The run_gen.py file was taken directly from the CodeT5 Git repository. The models.py file was also taken directly from the same repository. No updates were made to either file since the majority of the changes are going to be made in the utils.py and config.py files. 

**Screenshots that show the changes (if applicable):**

### CODE REVIEW

File: Fine Tuning/models.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+                 'bart': (BartConfig, BartForConditionalGeneration, BartTokenizer)}
+
+
+def get_model_size(model):
+    model_parameters = filter(lambda p: p.requires_grad, model.parameters())
+    model_size = sum([np.prod(p.size()) for p in model_parameters])
+    return "{}M".format(round(model_size / 1e+6))
+
+
+def build_or_load_gen_model(args):
```
Comment (oam67) [inline Fine Tuning/models.py:23]: Is this the function that creates the actual model object? I see model classes such as Seq2Seq, CloneModel, DefectModel being defined. Which model class would we specify for our project? This function is called in run_gen.py so it's great that you've included this script as well!

File: Fine Tuning/run_gen.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                loss = outputs.loss
+
+        eval_loss += loss.item()
+        batch_num += 1
+    eval_loss = eval_loss / batch_num
+    eval_ppl = round(np.exp(eval_loss), 5)
+    return eval_ppl
+
+
+def eval_bleu_epoch(args, eval_data, eval_examples, model, tokenizer, split_tag, criteria):
```
Comment (oam67) [inline Fine Tuning/run_gen.py:84]: I was wondering if we would still be doing evaluation with our fine-tuning? Since we only have training data we might have to evaluate the model on the data it has been trained on. Maybe we can look into creating a test/validation set later on but for now we can just focus on getting the model fine-tuned.

File: Fine Tuning/run_gen.py [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+    parser = argparse.ArgumentParser()
+    args = add_args(parser)
+    logger.info(args)
+    t0 = time.time()
+
+    set_dist(args)
+    set_seed(args)
+    config, model, tokenizer = build_or_load_gen_model(args)
+    model.to(args.device)
+    if args.n_gpu > 1:
```
Comment (oam67) [inline Fine Tuning/run_gen.py:179]: Does this mean we can run the model across multiple GPUs in parallel? For our use case would we just run the model on a single GPU? I know it can be slightly trickier when saving and loading models when we use multiple GPUs. I also don't think fine-tuning our model will be very time consuming or memory intensive.

Comment (AjithAkuthota23) [reply Fine Tuning/run_gen.py:179]: Is this part required for fine-tuning our model?

Comment (Omp06) [reply Fine Tuning/run_gen.py:179]: It might not be. I am currently changing the approach to fine-tuning the model.

**Review by AjithAkuthota23 — APPROVED** [review]
